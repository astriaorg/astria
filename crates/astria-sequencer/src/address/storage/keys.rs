pub(in crate::address) const BASE_PREFIX_KEY: &str = "address/prefixes/base";
pub(in crate::address) const IBC_COMPAT_PREFIX_KEY: &str = "address/prefixes/ibc_compat";

#[cfg(test)]
mod tests {
    use super::*;

    const COMPONENT_PREFIX: &str = "address/";

    #[test]
    fn keys_should_not_change() {
        insta::assert_snapshot!(BASE_PREFIX_KEY);
        insta::assert_snapshot!(IBC_COMPAT_PREFIX_KEY);
    }

    #[test]
    fn keys_should_have_component_prefix() {
        assert!(BASE_PREFIX_KEY.starts_with(COMPONENT_PREFIX));
        assert!(IBC_COMPAT_PREFIX_KEY.starts_with(COMPONENT_PREFIX));
    }
}
