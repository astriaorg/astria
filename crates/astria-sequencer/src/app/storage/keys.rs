pub(in crate::app) const CHAIN_ID: &str = "app/chain_id";
pub(in crate::app) const REVISION_NUMBER: &str = "app/revision_number";
pub(in crate::app) const BLOCK_HEIGHT: &str = "app/block_height";
pub(in crate::app) const BLOCK_TIMESTAMP: &str = "app/block_timestamp";
pub(in crate::app) const CONSENSUS_PARAMS: &str = "app/consensus_params";

pub(in crate::app) fn storage_version_by_height(height: u64) -> String {
    format!("app/storage_version/{height}")
}

#[cfg(test)]
mod tests {
    use super::*;

    const COMPONENT_PREFIX: &str = "app/";

    #[test]
    fn keys_should_not_change() {
        insta::assert_snapshot!("chain_id_key", CHAIN_ID);
        insta::assert_snapshot!("revision_number_key", REVISION_NUMBER);
        insta::assert_snapshot!("block_height_key", BLOCK_HEIGHT);
        insta::assert_snapshot!("block_timestamp_key", BLOCK_TIMESTAMP);
        insta::assert_snapshot!("consensus_params_key", CONSENSUS_PARAMS);
        insta::assert_snapshot!("storage_version_key", storage_version_by_height(42));
    }

    #[test]
    fn keys_should_have_component_prefix() {
        assert!(CHAIN_ID.starts_with(COMPONENT_PREFIX));
        assert!(REVISION_NUMBER.starts_with(COMPONENT_PREFIX));
        assert!(BLOCK_HEIGHT.starts_with(COMPONENT_PREFIX));
        assert!(BLOCK_TIMESTAMP.starts_with(COMPONENT_PREFIX));
        assert!(CONSENSUS_PARAMS.starts_with(COMPONENT_PREFIX));
        assert!(storage_version_by_height(42).starts_with(COMPONENT_PREFIX));
    }
}
