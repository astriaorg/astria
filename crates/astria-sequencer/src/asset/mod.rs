use std::sync::OnceLock;

use proto::native::sequencer::v1alpha1::asset::Denom;

pub(crate) static NATIVE_ASSET: OnceLock<Denom> = OnceLock::new();

pub(crate) fn initialize_native_asset(native_asset: &str) {
    if NATIVE_ASSET.get().is_some() {
        tracing::error!("native asset should only be set once");
        return;
    }

    let denom = Denom::from_base_denom(native_asset);
    NATIVE_ASSET
        .set(denom)
        .expect("native asset should only be set once");
}

pub(crate) fn get_native_asset() -> &'static Denom {
    NATIVE_ASSET.get().expect("native asset should be set")
}
