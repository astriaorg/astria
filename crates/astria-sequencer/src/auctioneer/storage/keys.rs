use crate::{
    accounts::AddressBytes,
    storage::keys::AccountPrefixer,
};

pub(in crate::auctioneer) const ENSHRINED_AUCTIONEER_PREFIX: &str = "auctioneer/";

pub(in crate::auctioneer) fn enshrined_auctioneer_key<T: AddressBytes>(address: &T) -> String {
    format!(
        "{}/auctioneer",
        AccountPrefixer::new(ENSHRINED_AUCTIONEER_PREFIX, address)
    )
}
