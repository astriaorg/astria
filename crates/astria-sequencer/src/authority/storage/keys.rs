use crate::accounts::AddressBytes;

pub(in crate::authority) const SUDO: &str = "authority/sudo";
pub(in crate::authority) const VALIDATOR_PREFIX: &str = "authority/validator/";
pub(in crate::authority) const VALIDATOR_COUNT: &str = "authority/validator_count";
pub(in crate::authority) const VALIDATOR_UPDATES: &str = "authority/validator_updates";

pub(in crate::authority) const PRE_ASPEN_VALIDATOR_SET: &str = "authority/validator_set"; // Deprecated post Aspen upgrade

pub(in crate::authority) fn validator<TAddress: AddressBytes>(address: &TAddress) -> String {
    format!(
        "{}/{}",
        VALIDATOR_PREFIX,
        hex::encode(address.address_bytes())
    )
}

#[cfg(test)]
mod tests {
    use astria_core::{
        crypto::ADDRESS_LENGTH,
        primitive::v1::Address,
    };

    use super::*;

    const COMPONENT_PREFIX: &str = "authority/";

    fn snapshot_validator() -> String {
        validator(
            &Address::builder()
                .array([1; ADDRESS_LENGTH])
                .prefix("astria")
                .try_build()
                .unwrap(),
        )
    }

    #[test]
    fn keys_should_not_change() {
        insta::assert_snapshot!("sudo_address_key", SUDO);
        insta::assert_snapshot!("validator_key", snapshot_validator());
        insta::assert_snapshot!("validator_count_key", VALIDATOR_COUNT);

        insta::assert_snapshot!("validator_set_key", PRE_ASPEN_VALIDATOR_SET);
        insta::assert_snapshot!("validator_updates_key", VALIDATOR_UPDATES);
    }

    #[test]
    fn keys_should_have_component_prefix() {
        assert!(SUDO.starts_with(COMPONENT_PREFIX));
        assert!(snapshot_validator().starts_with(COMPONENT_PREFIX));
        assert!(VALIDATOR_COUNT.starts_with(COMPONENT_PREFIX));

        assert!(PRE_ASPEN_VALIDATOR_SET.starts_with(COMPONENT_PREFIX));
        assert!(VALIDATOR_UPDATES.starts_with(COMPONENT_PREFIX));
    }
}
