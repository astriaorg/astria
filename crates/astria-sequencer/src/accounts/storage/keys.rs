use std::borrow::Cow;

use astria_core::primitive::v1::asset::IbcPrefixed;
use astria_eyre::eyre::{
    ContextCompat as _,
    Result,
    WrapErr as _,
};

use crate::{
    accounts::AddressBytes,
    storage::keys::{
        AccountPrefixer,
        Asset,
    },
};

pub(in crate::accounts) const TRANSFER_BASE_FEE: &str = "accounts/transfer_base_fee";
const COMPONENT_PREFIX: &str = "accounts/";
const BALANCE_PREFIX: &str = "balance/";
const NONCE: &str = "nonce";

/// Example: `accounts/gGhH....zZ4=/balance/`.
///                   |base64 chars|
pub(in crate::accounts) fn balance_prefix<TAddress: AddressBytes>(address: &TAddress) -> String {
    format!(
        "{}/{BALANCE_PREFIX}",
        AccountPrefixer::new(COMPONENT_PREFIX, address)
    )
}

/// Example: `accounts/gGhH....zZ4=/balance/0202....0202`.
///                   |base64 chars|       |64 hex chars|
pub(in crate::accounts) fn balance<'a, TAddress, TAsset>(
    address: &TAddress,
    asset: &'a TAsset,
) -> String
where
    TAddress: AddressBytes,
    &'a TAsset: Into<Cow<'a, IbcPrefixed>>,
{
    format!(
        "{}/{BALANCE_PREFIX}{}",
        AccountPrefixer::new(COMPONENT_PREFIX, address),
        Asset::from(asset)
    )
}

/// Example: `accounts/gGhH....zZ4=/nonce`.
///                   |base64 chars|
pub(in crate::accounts) fn nonce<TAddress: AddressBytes>(address: &TAddress) -> String {
    format!(
        "{}/{NONCE}",
        AccountPrefixer::new(COMPONENT_PREFIX, address)
    )
}

pub(in crate::accounts) fn extract_asset_from_key(key: &str) -> Result<IbcPrefixed> {
    Ok(key
        .strip_prefix(COMPONENT_PREFIX)
        .and_then(|s| s.split_once(BALANCE_PREFIX).map(|(_, asset)| asset))
        .wrap_err("failed to strip prefix from account balance key")?
        .parse::<Asset>()
        .wrap_err("failed to parse storage key suffix as address hunk")?
        .get())
}

#[cfg(test)]
mod tests {
    use astria_core::primitive::v1::{
        asset::Denom,
        Address,
    };

    use super::*;

    fn address() -> Address {
        "astria1rsxyjrcm255ds9euthjx6yc3vrjt9sxrm9cfgm"
            .parse()
            .unwrap()
    }

    fn asset() -> Denom {
        "an/asset/with/a/prefix".parse().unwrap()
    }

    #[test]
    fn keys_should_not_change() {
        insta::assert_snapshot!(TRANSFER_BASE_FEE);
        insta::assert_snapshot!(balance(&address(), &asset()));
        insta::assert_snapshot!(nonce(&address()));
    }

    #[test]
    fn keys_should_have_component_prefix() {
        assert!(TRANSFER_BASE_FEE.starts_with(COMPONENT_PREFIX));
        assert!(balance(&address(), &asset()).starts_with(COMPONENT_PREFIX));
        assert!(nonce(&address()).starts_with(COMPONENT_PREFIX));
    }

    #[test]
    fn balance_prefix_should_be_prefix_of_balance_key() {
        assert!(balance(&address(), &asset()).starts_with(&balance_prefix(&address())));
    }

    #[test]
    fn should_extract_asset_from_key() {
        let asset = IbcPrefixed::new([2; 32]);
        let key = balance(&[1; 20], &asset);
        let recovered_asset = extract_asset_from_key(&key).unwrap();
        assert_eq!(asset, recovered_asset);
    }
}
