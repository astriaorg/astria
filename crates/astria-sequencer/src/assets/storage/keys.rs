use std::borrow::Cow;

use astria_core::primitive::v1::asset::IbcPrefixed;
use astria_eyre::eyre::{
    bail,
    ContextCompat,
    Result,
    WrapErr,
};

use crate::storage::keys::Asset;

pub(in crate::assets) const NATIVE_ASSET: &str = "assets/native_asset";
pub(in crate::assets) const BLOCK_FEES_PREFIX: &[u8] = b"assets/block_fees/";
pub(in crate::assets) const FEE_ASSET_PREFIX: &[u8] = b"assets/fee_asset/";

/// Example: `assets/0101....0101`.
///                 |64 hex chars|
pub(in crate::assets) fn asset<'a, TAsset>(asset: &'a TAsset) -> String
where
    &'a TAsset: Into<Cow<'a, IbcPrefixed>>,
{
    format!("assets/{}", Asset::from(asset))
}

pub(in crate::assets) fn fee_asset<'a, TAsset>(asset: &'a TAsset) -> Vec<u8>
where
    &'a TAsset: Into<Cow<'a, IbcPrefixed>>,
{
    [FEE_ASSET_PREFIX, Asset::from(asset).as_bytes()].concat()
}

pub(in crate::assets) fn block_fees<'a, TAsset>(asset: &'a TAsset) -> Vec<u8>
where
    &'a TAsset: Into<Cow<'a, IbcPrefixed>>,
{
    [BLOCK_FEES_PREFIX, Asset::from(asset).as_bytes()].concat()
}

pub(in crate::assets) fn extract_asset_from_fee_asset_key(key: &[u8]) -> Result<IbcPrefixed> {
    extract_asset_from_key(key, FEE_ASSET_PREFIX)
        .wrap_err("failed to extract asset from fee asset key")
}

pub(in crate::assets) fn extract_asset_from_block_fees_key(key: &[u8]) -> Result<IbcPrefixed> {
    extract_asset_from_key(key, BLOCK_FEES_PREFIX)
        .wrap_err("failed to extract asset from fee asset key")
}

fn extract_asset_from_key(key: &[u8], prefix: &[u8]) -> Result<IbcPrefixed> {
    let suffix = key.strip_prefix(prefix).wrap_err_with(|| {
        format!(
            "key `{}` did not have prefix `{}`",
            telemetry::display::hex(key),
            telemetry::display::hex(prefix)
        )
    })?;
    if suffix.len() != IbcPrefixed::ENCODED_HASH_LEN {
        bail!(
            "suffix `{}` of key `{}` is not {} bytes",
            telemetry::display::hex(suffix),
            telemetry::display::hex(key),
            IbcPrefixed::ENCODED_HASH_LEN
        );
    }
    let mut buffer = [0; IbcPrefixed::ENCODED_HASH_LEN];
    buffer.copy_from_slice(suffix);
    Ok(IbcPrefixed::new(buffer))
}

#[cfg(test)]
mod tests {
    use astria_core::primitive::v1::asset::Denom;
    use telemetry::display::base64;

    use super::*;

    const COMPONENT_PREFIX: &str = "assets/";

    fn test_asset() -> Denom {
        "an/asset/with/a/prefix".parse().unwrap()
    }

    #[test]
    fn keys_should_not_change() {
        insta::assert_snapshot!(NATIVE_ASSET);
        insta::assert_snapshot!(asset(&test_asset()));
        insta::assert_snapshot!(base64(&fee_asset(&test_asset())));
        insta::assert_snapshot!(base64(&block_fees(&test_asset())));
    }

    #[test]
    fn keys_should_have_component_prefix() {
        assert!(NATIVE_ASSET.starts_with(COMPONENT_PREFIX));
        assert!(asset(&test_asset()).starts_with(COMPONENT_PREFIX));
        assert!(fee_asset(&test_asset()).starts_with(COMPONENT_PREFIX.as_bytes()));
        assert!(block_fees(&test_asset()).starts_with(COMPONENT_PREFIX.as_bytes()));
    }

    #[test]
    fn prefixes_should_be_prefixes_of_relevant_keys() {
        assert!(fee_asset(&test_asset()).starts_with(FEE_ASSET_PREFIX));
        assert!(block_fees(&test_asset()).starts_with(BLOCK_FEES_PREFIX));
    }

    #[test]
    fn should_extract_asset_from_key() {
        let asset = IbcPrefixed::new([1; 32]);

        let key = fee_asset(&asset);
        let recovered_asset = extract_asset_from_fee_asset_key(&key).unwrap();
        assert_eq!(asset, recovered_asset);

        let key = block_fees(&asset);
        let recovered_asset = extract_asset_from_block_fees_key(&key).unwrap();
        assert_eq!(asset, recovered_asset);
    }
}
