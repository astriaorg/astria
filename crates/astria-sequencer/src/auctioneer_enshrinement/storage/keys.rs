use crate::accounts::AddressBytes;
use crate::storage::keys::AccountPrefixer;

pub(in crate::auctioneer_enshrinement) const ENSHRINED_AUCTIONEER_PREFIX: &str = "auctioneer_enshrinement/";

pub(in crate::auctioneer_enshrinement) fn enshrined_auctioneer_key<T: AddressBytes>(address: &T) -> String {
    format!(
        "{}/auctioneer",
        AccountPrefixer::new(ENSHRINED_AUCTIONEER_PREFIX, address)
    )
}