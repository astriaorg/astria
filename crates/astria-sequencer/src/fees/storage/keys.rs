use std::borrow::Cow;

use astria_core::primitive::v1::asset::IbcPrefixed;
use astria_eyre::eyre::{
    self,
    eyre,
    Context as _,
};

use crate::storage::keys::Asset;

pub(in crate::fees) const TRANSFER: &str = "fees/transfer";
pub(in crate::fees) const SEQUENCE: &str = "fees/sequence";
pub(in crate::fees) const ICS20_WITHDRAWAL: &str = "fees/ics20_withdrawal";
pub(in crate::fees) const INIT_BRIDGE_ACCOUNT: &str = "fees/init_bridge_account";
pub(in crate::fees) const BRIDGE_LOCK: &str = "fees/bridge_lock";
pub(in crate::fees) const BRIDGE_UNLOCK: &str = "fees/bridge_unlock";
pub(in crate::fees) const BRIDGE_SUDO_CHANGE: &str = "fees/bridge_sudo_change";
pub(in crate::fees) const IBC_RELAY: &str = "fees/ibc_relay";
pub(in crate::fees) const VALIDATOR_UPDATE: &str = "fees/validator_update";
pub(in crate::fees) const FEE_ASSET_CHANGE: &str = "fees/fee_asset_change";
pub(in crate::fees) const FEE_CHANGE: &str = "fees/fee_change";
pub(in crate::fees) const IBC_RELAYER_CHANGE: &str = "fees/ibc_relayer_change";
pub(in crate::fees) const SUDO_ADDRESS_CHANGE: &str = "fees/sudo_address_change";
pub(in crate::fees) const IBC_SUDO_CHANGE: &str = "fees/ibc_sudo_change";
pub(in crate::fees) const BLOCK: &str = "fees/block"; // NOTE: `BLOCK` is only used in the ephemeral store.
pub(in crate::fees) const ALLOWED_ASSET_PREFIX: &str = "fees/allowed_asset/";

pub(in crate::fees) fn allowed_asset<'a, TAsset>(asset: &'a TAsset) -> String
where
    &'a TAsset: Into<Cow<'a, IbcPrefixed>>,
{
    format!("{ALLOWED_ASSET_PREFIX}{}", Asset::from(asset))
}

pub(in crate::fees) fn extract_asset_from_allowed_asset_key(
    key: &str,
) -> eyre::Result<IbcPrefixed> {
    extract_asset_from_key(key, ALLOWED_ASSET_PREFIX)
        .wrap_err("failed to extract asset from fee asset key")
}

fn extract_asset_from_key(key: &str, prefix: &str) -> eyre::Result<IbcPrefixed> {
    let suffix = key
        .strip_prefix(prefix)
        .ok_or_else(|| eyre!("key `{key}` did not have prefix `{prefix}`"))?;
    suffix.parse().wrap_err_with(|| {
        format!("failed to parse suffix `{suffix}` of key `{key}` as an ibc-prefixed asset",)
    })
}

#[cfg(test)]
mod tests {
    use astria_core::primitive::v1::asset::Denom;
    use insta::assert_snapshot;

    use super::*;

    const COMPONENT_PREFIX: &str = "fees/";

    fn test_asset() -> Denom {
        "an/asset/with/a/prefix".parse().unwrap()
    }

    #[test]
    fn keys_should_not_change() {
        // NOTE: This helper struct is just to avoid having 14 snapshot files to contend with.
        // NOTE: `BLOCK` is only used in the ephemeral store, so isn't included here.
        assert_snapshot!("bridge_lock_fees_key", BRIDGE_LOCK);
        assert_snapshot!("bridge_sudo_change_fees_key", BRIDGE_SUDO_CHANGE);
        assert_snapshot!("bridge_unlock_fees_key", BRIDGE_UNLOCK);
        assert_snapshot!("fee_asset_change_fees_key", FEE_ASSET_CHANGE);
        assert_snapshot!("allowed_asset_prefix", ALLOWED_ASSET_PREFIX);
        assert_snapshot!("fee_change_fees_key", FEE_CHANGE);
        assert_snapshot!("ibc_relay_fees_key", IBC_RELAY);
        assert_snapshot!("ibc_relayer_change_fees_key", IBC_RELAYER_CHANGE);
        assert_snapshot!("ibc_sudo_change_fees_key", IBC_SUDO_CHANGE);
        assert_snapshot!("ics20_withdrawal_fees_key", ICS20_WITHDRAWAL);
        assert_snapshot!("init_bridge_account_fees_key", INIT_BRIDGE_ACCOUNT);
        assert_snapshot!("sequence_fees_key", SEQUENCE);
        assert_snapshot!("sudo_address_change_fees_key", SUDO_ADDRESS_CHANGE);
        assert_snapshot!("transer_fees_key", TRANSFER);
        assert_snapshot!("validator_update_fees_key", VALIDATOR_UPDATE);
        assert_snapshot!("allowed_asset_key", allowed_asset(&test_asset()));
    }

    #[test]
    fn keys_should_have_component_prefix() {
        assert!(TRANSFER.starts_with(COMPONENT_PREFIX));
        assert!(SEQUENCE.starts_with(COMPONENT_PREFIX));
        assert!(ICS20_WITHDRAWAL.starts_with(COMPONENT_PREFIX));
        assert!(INIT_BRIDGE_ACCOUNT.starts_with(COMPONENT_PREFIX));
        assert!(BRIDGE_LOCK.starts_with(COMPONENT_PREFIX));
        assert!(BRIDGE_UNLOCK.starts_with(COMPONENT_PREFIX));
        assert!(BRIDGE_SUDO_CHANGE.starts_with(COMPONENT_PREFIX));
        assert!(IBC_RELAY.starts_with(COMPONENT_PREFIX));
        assert!(VALIDATOR_UPDATE.starts_with(COMPONENT_PREFIX));
        assert!(FEE_ASSET_CHANGE.starts_with(COMPONENT_PREFIX));
        assert!(FEE_CHANGE.starts_with(COMPONENT_PREFIX));
        assert!(IBC_RELAYER_CHANGE.starts_with(COMPONENT_PREFIX));
        assert!(SUDO_ADDRESS_CHANGE.starts_with(COMPONENT_PREFIX));
        assert!(IBC_SUDO_CHANGE.starts_with(COMPONENT_PREFIX));
        assert!(ALLOWED_ASSET_PREFIX.starts_with(COMPONENT_PREFIX));
        assert!(allowed_asset(&test_asset()).starts_with(COMPONENT_PREFIX));
    }

    #[test]
    fn prefixes_should_be_prefixes_of_relevant_keys() {
        assert!(allowed_asset(&test_asset()).starts_with(ALLOWED_ASSET_PREFIX));
    }

    #[test]
    fn should_extract_asset_from_key() {
        let asset = IbcPrefixed::new([1; 32]);

        let key = allowed_asset(&asset);
        let recovered_asset = extract_asset_from_allowed_asset_key(&key).unwrap();
        assert_eq!(asset, recovered_asset);
    }
}
