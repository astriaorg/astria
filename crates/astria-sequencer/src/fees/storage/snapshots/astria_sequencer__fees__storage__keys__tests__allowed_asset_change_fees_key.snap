---
source: crates/astria-sequencer/src/fees/storage/keys.rs
expression: FEE_ASSET_CHANGE
---
fees/allowed_asset_change
